### PR TITLE
Drag and drop

### PR DESCRIPTION
**Pull request by ItzhakIsraeli**

### CODE REVIEW

File: client/src/components/ClickableBox.tsx [hunk 1]
```diff
@@ -34,14 +34,14 @@ export const ClickableBox = ({
 
     return (
         <View style={{marginVertical: 10, marginHorizontal: 10}}>
-            <RelativeView>
+            <RelativeView style={{backgroundColor: '#f2f2f2'}}>
                 {hasIcon &&
                     <>
                         <TextToSpeechIcon text={name}
                                           gender={userGender as Gender}
                                           style={{position: 'absolute', top: 3, left: 6, zIndex: 1}}/>
```
Comment (AviranBac) [inline client/src/components/ClickableBox.tsx:42]: Add backgroundColor: '#f2f2f2' to TextToSpeechIcon too
Also add backgroundColor to FavoriteIcon (look at the comment in FavoriteIcon where you should add it)

File: client/src/components/icons/FavoriteIcon.tsx [hunk 2]
```diff
@@ -49,7 +49,7 @@ export const FavoriteIcon = ({itemId}: Props) => {
                 <View style={IconsStyles.iconCircleWrapper}>
                     <View style={IconsStyles.iconCirclePaper}/>
```
Comment (AviranBac) [inline client/src/components/icons/FavoriteIcon.tsx:50]: Add optional backgroundColor props (default: white) and do:
`<View style={IconsStyles.iconCirclePaper}/>` -> `<View style={{...IconsStyles.iconCirclePaper, backgroundColor}}/>`

Remove backgroundColor from iconCirclePaper object so it won't confuse us

**Review by AviranBac — CHANGES_REQUESTED** [review]

File: client/src/components/DragAndDrop/styles.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        borderWidth: 1,
+        borderColor: '#ff6548',
+        marginRight: 15,
+        borderRadius: 12
+    },
+    hurdle_edit_text: {
+        color: '#ff6548',
+        fontSize: 16
+    },
+    // selected_container: {
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/src/components/DragAndDrop/styles.ts:37]: done

File: client/src/services/items.service.ts [hunk 4]
```diff
@@ -20,4 +20,11 @@ export const recordItemChosen = (item: Item, recommendedItemIds?: string[]) => {
         ...(recommendedItemIds && {recommendedItemIds})
     })
         .then((response) => response.data);
+}
+
+export const updateItemListOrder = (orderedItemIds: string[], categoryId?: string) => {
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/src/services/items.service.ts:25]: done

File: client/.gitignore [hunk 5]
```diff
@@ -1,6 +1,7 @@
 node_modules/
 .expo/
 dist/
+package-lock.json
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/.gitignore:4]: done

File: client/src/components/DragAndDrop/DragAndDrop.tsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            </View>
+        )
+    }
+
+    const onSelectedDragEnd = () => {
+        setScrollEnabled(true);
+    }
+
+    const onSelectedDragStart = () => {
+        if (!isEditState) {
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/src/components/DragAndDrop/DragAndDrop.tsx:71]: done

File: client/src/components/DragAndDrop/DragAndDrop.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    items: any[];
+    onItemPress?: (item: any) => void;
+    category?: Category;
+    isFavorites?: boolean;
+    isHomeScreen?: boolean;
+}
+
+export const DragAndDrop = ({items, onItemPress, category, isFavorites = false, isHomeScreen = false}: Props) => {
+    const [scrollEnabled, setScrollEnabled] = useState(true);
+    const [isEditState, setIsEdit] = useState(false);
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/src/components/DragAndDrop/DragAndDrop.tsx:29]: done

File: client/src/components/DragAndDrop/DragAndDrop.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        setIsEdit(false);
+        const sortedMap = new Map(Array.from(setMap.current.entries()).sort((a, b) => a[1] - b[1]));
+        const categoryId = isFavorites ? 'favorites' : category?._id;
+        Toast.show({
+            type: 'success',
+            text1: 'שמירה בוצעה בהצלחה',
+            text2: 'סדר הפריטים במערכת ⭐️',
+        });
+
+        isFavorites ? updateFavoriteListOrder(Array.from(sortedMap.keys())).then(setItemsList)
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/src/components/DragAndDrop/DragAndDrop.tsx:89]: done

File: client/src/components/DragAndDrop/DragAndDrop.tsx [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        } else {
+            setScrollEnabled(false);
+        }
+    }
+
+    const onSaveChanges = () => {
+        setIsEdit(false);
+        const sortedMap = new Map(Array.from(setMap.current.entries()).sort((a, b) => a[1] - b[1]));
+        const categoryId = isFavorites ? 'favorites' : category?._id;
+        Toast.show({
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/src/components/DragAndDrop/DragAndDrop.tsx:83]: done

File: client/src/components/DragAndDrop/DragAndDrop.tsx [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+                    </TouchableOpacity>
+                    <TouchableOpacity style={styles.hurdle_edit} onPress={handleReset}>
+                        <Text style={styles.hurdle_edit_text}>{'ביטול'}</Text>
+                    </TouchableOpacity>
+                </View>}
+                <ItemsWrapper>
+                    <DragSortableView
+                        dataSource={itemsList}
+                        parentWidth={deviceWidth}
+                        childrenWidth={isPortrait ? childrenWidth : childrenWidth}
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/src/components/DragAndDrop/DragAndDrop.tsx:115]: as I told you in the whatsApp

File: client/src/components/ClickableBox.tsx (hunk 1 above)
Comment (ItzhakIsraeli) [reply client/src/components/ClickableBox.tsx:42]: done

File: client/src/components/icons/FavoriteIcon.tsx (hunk 2 above)
Comment (ItzhakIsraeli) [reply client/src/components/icons/FavoriteIcon.tsx:50]: done

File: client/src/components/DragAndDrop/DragAndDrop.tsx [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    };
+
+    return (
+        <SafeAreaView style={{backgroundColor: '#fff', flex: 1}}>
+            <ScrollView
+                scrollEnabled={scrollEnabled}>
+                {isEdit && <View style={styles.hurdle}>
+                    <StyledButton title={'שמירת שינויים'} onPress={onSaveChanges}/>
+                    <StyledButton title={'ביטול'} onPress={handleReset}/>
+                    {/*<TouchableOpacity style={styles.hurdle_edit} onPress={onSaveChanges}>*/}
```
Comment (AviranBac) [inline client/src/components/DragAndDrop/DragAndDrop.tsx:111]: ?

**Review by AviranBac — CHANGES_REQUESTED** [review]

File: client/src/components/DragAndDrop/DragAndDrop.tsx [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        const array: string[] = Array.from(iterable);
+
+        updateOrderFunc(array).then((res) => {
+            setItemsList(res);
+            Toast.show({
+                type: 'success',
+                text1: 'שמירה בוצעה בהצלחה',
+                text2: 'סדר הפריטים במערכת עודכן ⭐️',
+            });
+        }).catch(() => Toast.show({
```
Comment (AviranBac) [inline client/src/components/DragAndDrop/DragAndDrop.tsx:91]: Does it work?

**Review by AviranBac — CHANGES_REQUESTED** [review]

File: client/src/components/DragAndDrop/DragAndDrop.tsx [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+const deviceWidth = Dimensions.get('window').width
+const deviceHeight = Dimensions.get('window').height
+
+const childrenWidth = deviceWidth / 4;
+const childrenHeight = deviceHeight / 4;
+
+interface Props {
+    items: any[];
+    onItemPress?: (item: any) => void;
+    isHomeScreen?: boolean;
```
[the remark this answers is not in the document]
Comment (ItzhakIsraeli) [reply client/src/components/DragAndDrop/DragAndDrop.tsx:19]: done

File: client/src/components/DragAndDrop/DragAndDrop.tsx (hunk 11 above)
Comment (ItzhakIsraeli) [reply client/src/components/DragAndDrop/DragAndDrop.tsx:111]: done

**Review by AviranBac — APPROVED** [review]

**Pull request merged**